Dev

sseifeln opened the pull request:
Input router synthesizes and compiles, but not yet verified against emulation results.

pwittich commented:
Does this supersede  #63 ?
And we currently don't want to store many data files in the repo (I think you add a large number); we want to add just the minimum and download the rest from CERN box. See e.g. the way @aehart downloads his files see [this example](https://github.com/cms-tracklet/firmware-hls/blob/master/emData/TC/download.sh ) and top-level readme.

sseifeln commented:
Yes it does - sorry probably should have pushed back to my master branch and 'resurrected' #63 rather than starting a new one .. got too excited about it synthesizing! 
I've cleaned up the test bench so it only used the file that is already in the repo by default. I've also added the script (thanks!) to download other files from my /eos/ if needed.

sseifeln commented on TrackletAlgorithm/InputRouterTop.cpp:
```diff
@@ -1,143 +1,248 @@
+	ap_uint<kNBits_DTC> hInputStub;
+	LOOP_InputStubsPS : for (int cStubCounter=0; cStubCounter<kMaxStubsFromLink; cStubCounter++)
 	{
 		#pragma HLS pipeline II=1
```
Just found the ruler option in Sublime .. so yes will do!

sseifeln replied:
Ok - done!

djcranshaw commented on TrackletAlgorithm/Constants.hh:
```diff
@@ -19,7 +24,6 @@ constexpr int kNBits_BX = 3;
 constexpr int kNBits_MemAddrBinned = 4;
 constexpr int kMemDepthBinned = (1<<kNBits_MemAddrBinned);
 
-constexpr int kNBits_z = 3;
```
Did you get rid of this const on purpose? I think the MatchEngine uses it.

sseifeln replied:
Nope not on purpose - sorry! Back now ..

djcranshaw commented:
So it looks like all the changes to TrackletAlgorithm/MemoryTemplate.hh are white-space-related, with no functional changes. Is this true and if so, can we just leave it as it so that all other branches don't have diverging histories for this file?

sseifeln commented:
Ok - did that. Also just realized that I did not have one file tracked in git (TrackletAlgorithm/InputRouter.cc) - so have also done that.. sorry about that!

djcranshaw commented:
So the changes to InputStubMemory.hh (defining InputRouterMemory which inherits from MemoryTemplate and then type alias-ing InputStubMemory as InputRouterMemory) I think will cause different ports to be synthesized relative to the rest of the project, which will wreak havoc on the project generation scripts. It's also a significant change to how the rest of the memory modules work in the rest of the project. Here I mostly mean StubsBarrelPS hBrlPS, StubsDiskPS hDskPS, StubsBarrel2S hBrl2S, StubsDisk2S hDsk2S, which are the outputs of IR. The inputs to IR are streamed from the DTCs (and whatever mapping ROMs you may need) so these will of course have to look different from our between-block BRAMs. Before we merge, I'd like to have a look and see if we can make some minimal changes which will allow us to retain the memory structure for InputStubMemory that we have for the rest of the memories, without having to significantly change the logic of IR. Unfortunately I've run out of time for that today but I'll look at this tomorrow.

djcranshaw commented:
One other thing, in InputStubMemory.hh you define a DTCMap class and a DTCMapMemory type alias, but these are nowhere used in the rest of the IR as far as I can see. Can these classes be removed?

sseifeln commented:
I checked and the IR will run with the declaration of InputStubMemory used in the master branch... so in case it helps at all I've : 
1) uncommented the type declaration in InputStubMemory.hh - sorry should have checked that I wasn't modifying a class used elsewhere (VMRouter)
2) for now have changed the name of the new definition I created ...  if you don't think its compatible with the project I'll simply remove it

I've also cleaned up the DTCMap classes :)

djcranshaw commented:
So Sarah and I have decided to hold off on this PR until the the interface generated by the top-level matches what the project_generation scripts and BRAMs implemented by memory.v expect, and until it passes CSim. Also we'll close the PR so that there isn't one sitting here apparently waiting for a review (can be reopened if anyone disagrees).
